### PR TITLE
feat(playground): projects + RAG knowledge base

### PR DESCRIPTION
**Pull request by smakosh**
## Overview

Adds **Projects** to the chat playground (chat.llmgateway.io) — the same mental model as Claude's Projects: group related chats under a project, upload files as a **knowledge base**, and have every chat in the project answer **grounded in those files via RAG**, with per-project custom instructions.

### Playground UI
- **Projects** nav item + `/projects` page: create/edit/delete projects (name, description, instructions), sidebar list with file/chat counts, org-context aware (`?orgId=` preserved everywhere, same semantics as chats)
- **Knowledge base panel**: upload text/markdown/code files (client reads as text), live *Chunking and embedding…* state, `Indexed` badge with chunk counts, per-file delete, error state surfaced on the file row
- **Chats in this project** list + *New chat* button that opens the chat page in project context
- Chat page shows a **project context banner** (name + file count, links back to the project) and associates new chats with the project (`chat.projectId`), including 404-recovery re-creation and fork inheritance

### RAG pipeline
- Upload: text is chunked (~1500 chars, paragraph-aware with overlap for oversized paragraphs) and embedded through the gateway's own `/v1/embeddings` (`openai/text-embedding-3-small`), stored per-chunk
- Retrieval: `POST /chat-projects/{id}/retrieve` embeds the query and returns top-k chunks by cosine similarity (score-thresholded)
- The playground chat route calls retrieve server-side before streaming and injects project instructions + retrieved excerpts (with source file names) into the system prompt; the model cites the files it used
- Billing stays consistent: the query embedding is billed to the **same gateway key as the chat request** (passed via `x-llmgateway-key`); browser-initiated uploads bill via the playground key cookie
- Embeddings are stored as `jsonb` and scored in Node — **no pgvector / infra change needed** at knowledge-base scale (capped at 20 files / 500k chars per file)

### API (`apps/api`)
- New `/chat-projects` routes: CRUD, file upload/list/delete, retrieve — user-scoped ownership checks throughout
- `chats` routes gain `projectId`: create (validated), list (optional `projectId` filter), search/get/update responses, fork inheritance
- Playground key resolution (cookie → chat-org default key fallback) extracted from `skills.ts` into `utils/playground-key.ts` and reused

### DB (`packages/db`)
- `chat_project`, `chat_project_file`, `chat_project_file_chunk` (+ relations) and `chat.project_id` (`SET NULL` on project delete — chats survive, they just lose project context)
- Migration generated with `pnpm migrations`

### Blog post
- Announcement post at `/blog/chat-projects-knowledge-base` (`apps/ui/src/content/blog/2026-07-05-chat-projects-knowledge-base.md`) with a gpt-image-2 OG image + composited logo

## Testing
- `rag.spec.ts`: chunking edge cases (empty, grouping, hard-split overlap, CRLF) + cosine similarity properties — 11 tests
- `chat-projects.spec.ts`: CRUD, cross-user 403, empty-content 400, retrieve empty-KB short-circuit, chats projectId create/filter — 8 tests
- Full `pnpm test:unit`: **141 files, 2365 passed**
- Full `pnpm build` green
- Verified end-to-end locally (curl + browser): upload → chunks embedded → retrieve scores correctly → chat answer grounded in file facts with `(nova-handbook.md)` citations, using `openai/gpt-5-mini` under a dashboard-org context; also verified the insufficient-credits error path marks the file row `error`

## Demo
17s demo video recorded (attached via web UI drag-drop — GitHub blocks video upload from CLI).

## Notes
- Binary formats (PDF) are not accepted yet — files are read as text client-side; wiring the existing OCR route in as an extraction step is a natural follow-up
- Retrieval failure degrades gracefully to a normal chat rather than failing the message

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01N77ENja1mk23QGwkHyF5EQ

## Summary by CodeRabbit

* **New Features**
  * Added Projects for chats, with a sidebar entry and a dedicated Projects page.
  * Users can create, edit, delete, and organize chat projects, attach files, and view related chats.
  * Chats can now be linked to a project, and project context can be shown in the chat UI.
  * Project knowledge bases now power context-aware retrieval for chat responses.

* **Bug Fixes**
  * Improved validation and error handling for project files, project access, and retrieval behavior.

### CODE REVIEW

File: apps/api/src/utils/playground-key.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+			.insert(tables.apiKey)
+			.values({
+				token: prefix + shortid(40),
+				projectId: project.id,
+				description: "Auto-generated playground key",
+				usageLimit: null,
+				createdBy: user.id,
+			})
+			.returning();
+	}
```
[the remark this answers is not in the document]
Comment (smakosh) [reply apps/api/src/utils/playground-key.ts:74]: This create-or-reuse block is pre-existing behavior extracted verbatim from skills.ts (see the refactor in this PR) — not new logic. The race worst case is a duplicate default project/key on the user's own Chat org, both of which remain valid, so making it atomic (advisory lock or unique constraint) is out of scope here. Happy to tackle repo-wide in a follow-up since skills generation and the ensure-key flow share the same pattern.

File: apps/playground/src/app/api/chat/route.ts [hunk 2]
```diff
@@ -858,6 +881,53 @@ export async function POST(req: Request) {
...
+						.map((chunk) => `[Source: ${chunk.fileName}]\n${chunk.content}`)
+						.join("\n\n---\n\n")}`,
+				);
+			}
+			if (sections.length) {
+				projectContext = `You are answering inside the project "${retrieval.project.name}".\n\n${sections.join("\n\n")}`;
+			}
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (smakosh) [reply apps/playground/src/app/api/chat/route.ts:933]: Added AbortSignal.timeout(15_000) to the retrieval call in 450cecf — on timeout fetchServerData returns null and the chat proceeds without project context.

File: apps/playground/src/components/playground/chat-page-client.tsx [hunk 3]
```diff
@@ -902,6 +914,18 @@ export default function ChatPageClient({
...
+	useEffect(() => {
+		if (currentChatId) {
+			if (currentChatData?.chat?.id === currentChatId) {
+				setActiveProjectId(currentChatData.chat.projectId ?? null);
+			}
+		} else {
+			setActiveProjectId(projectIdFromUrl);
+		}
+	}, [currentChatId, currentChatData?.chat, projectIdFromUrl]);
+
```
[the remark this answers is not in the document]
Comment (smakosh) [reply apps/playground/src/components/playground/chat-page-client.tsx:933]: Fixed in 450cecf: when navigating to a different chat whose detail hasn't loaded yet, activeProjectId is cleared (guarded by pendingNewChatRef so a chat just created from ?project= keeps its context for the first send).

File: apps/playground/src/components/playground/chat-page-client.tsx [hunk 4]
```diff
@@ -1974,6 +1999,9 @@ export default function ChatPageClient({
 							previewPrompt={getFirstUserMessageText(messages)}
 						/>
 					</header>
+					{activeProjectId && !isTemporaryChat ? (
+						<ProjectContextBanner projectId={activeProjectId} />
+					) : null}
```
[the remark this answers is not in the document]
Comment (smakosh) [reply apps/playground/src/components/playground/chat-page-client.tsx:2009]: Fixed in 450cecf: the banner now reads ?orgId= via useSearchParams and appends it to the /projects link.

File: apps/playground/src/hooks/useChatProjects.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	return api.useMutation("delete", "/chat-projects/{id}", {
+		onSuccess: () => {
+			invalidate();
+			toast("Project deleted");
+		},
+		onError: (error) => {
+			toast.error(getErrorMessage(error));
+		},
+	});
+}
```
[the remark this answers is not in the document]
Comment (smakosh) [reply apps/playground/src/hooks/useChatProjects.ts:119]: Fixed in 450cecf with the suggested change — delete now invalidates the project's detail query too.

File: packages/db/src/schema.ts [hunk 6]
```diff
@@ -1743,6 +1825,10 @@ export const chat = pgTable(
 		parentChatId: text().references((): AnyPgColumn => chat.id, {
 			onDelete: "cascade",
 		}),
+		// Chat project (knowledge base) this chat belongs to, if any.
+		projectId: text().references(() => chatProject.id, {
+			onDelete: "set null",
+		}),
```
[the remark this answers is not in the document]
Comment (smakosh) [reply packages/db/src/schema.ts:1831]: Added chat_project_id_idx in 450cecf via a regenerated migration (1783257466_silly_lockjaw.sql).

**Comment by steebchen** [comment]
Images automagically compressed by [Calibre](https://calibreapp.com)'s [image-actions](https://github.com/marketplace/actions/image-actions) ✨

Compression reduced images by <strong>68.7%</strong>, saving <strong>1.2 MB</strong>.

| Filename | Before | After | Improvement | Visual comparison |
| -------- | ------ | ----- | ----------- | ----------------- |
| <code>apps/ui/public/blog/chat-projects-knowledge-base.png</code> | 1.8 MB | 569.2 KB | 68.7% | [View diff](/theopenco/llmgateway/pull/2921/commits/603f9b0584b6f73ea79ac8873bff7b0d46fec2d3?short_path=8843ce8#diff-8843ce8d73585da415226dfec30cad91) |

**Comment by steebchen** [comment]
Images automagically compressed by [Calibre](https://calibreapp.com)'s [image-actions](https://github.com/marketplace/actions/image-actions) ✨

Compression reduced images by <strong>11.2%</strong>, saving <strong>63.6 KB</strong>.

| Filename | Before | After | Improvement | Visual comparison |
| -------- | ------ | ----- | ----------- | ----------------- |
| <code>apps/ui/public/blog/chat-projects-knowledge-base.png</code> | 569.2 KB | 505.6 KB | 11.2% | [View diff](/theopenco/llmgateway/pull/2921/commits/3aaccd63c7402cee5248ae9edf37e10ef8828a4d?short_path=8843ce8#diff-8843ce8d73585da415226dfec30cad91) |

**Comment by steebchen** [comment]
Images automagically compressed by [Calibre](https://calibreapp.com)'s [image-actions](https://github.com/marketplace/actions/image-actions) ✨

Compression reduced images by <strong>10.8%</strong>, saving <strong>54.5 KB</strong>.

| Filename | Before | After | Improvement | Visual comparison |
| -------- | ------ | ----- | ----------- | ----------------- |
| <code>apps/ui/public/blog/chat-projects-knowledge-base.png</code> | 505.6 KB | 451.0 KB | 10.8% | [View diff](/theopenco/llmgateway/pull/2921/commits/041767ab9b98ee8bef519fcae1e40f652d1cbd79?short_path=8843ce8#diff-8843ce8d73585da415226dfec30cad91) |

**Comment by steebchen** [comment]
Images automagically compressed by [Calibre](https://calibreapp.com)'s [image-actions](https://github.com/marketplace/actions/image-actions) ✨

Compression reduced images by <strong>7.6%</strong>, saving <strong>34.5 KB</strong>.

| Filename | Before | After | Improvement | Visual comparison |
| -------- | ------ | ----- | ----------- | ----------------- |
| <code>apps/ui/public/blog/chat-projects-knowledge-base.png</code> | 451.0 KB | 416.5 KB | 7.6% | [View diff](/theopenco/llmgateway/pull/2921/commits/3d99162bbd76d23ad8dc275c8955c963c511843d?short_path=8843ce8#diff-8843ce8d73585da415226dfec30cad91) |

**Comment by steebchen** [comment]
Images automagically compressed by [Calibre](https://calibreapp.com)'s [image-actions](https://github.com/marketplace/actions/image-actions) ✨

Compression reduced images by <strong>13.7%</strong>, saving <strong>57.2 KB</strong>.

| Filename | Before | After | Improvement | Visual comparison |
| -------- | ------ | ----- | ----------- | ----------------- |
| <code>apps/ui/public/blog/chat-projects-knowledge-base.png</code> | 416.5 KB | 359.3 KB | 13.7% | [View diff](/theopenco/llmgateway/pull/2921/commits/3dee390c43117de3b49241bbde93874e7f979108?short_path=8843ce8#diff-8843ce8d73585da415226dfec30cad91) |
